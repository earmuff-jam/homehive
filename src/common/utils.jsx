/**
 * Utils file
 *
 * Used to create common minor functions that can be re-used across the application
 */

import { Typography } from "@mui/material";

/**
 * pluralize
 *
 * function used to add a plural form where applicable.
 *
 * @param {int} arrLength - the length of the array that needs to be tabulated.
 * @param {string} wordStr - the string representation of the selected word.
 *
 * @returns a plural form of the selected word
 *
 */
export function pluralize(arrLength, wordStr) {
  if (arrLength <= 1) return wordStr;
  return `${wordStr}s`;
}

/**
 * createHelperSentences
 *
 * function used to create helper sentences for tour steps.
 * @param {string} verbStr - the string that replaces the verb in each of the sentences
 * @param {string} extraClauseStr - the string that replaces the noun or action in each of the sentences
 */
export function createHelperSentences(verbStr, extraClauseStr) {
  return (
    <Typography variant="caption">
      This help / guide is designed to aide you in learning how to{" "}
      {verbStr + extraClauseStr}? Feel free to restart the guide if necessary.
    </Typography>
  );
}