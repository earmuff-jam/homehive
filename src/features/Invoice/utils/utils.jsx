import dayjs from "dayjs";

function populateMap(items = [], columnName, uniqueMap) {
  items.forEach((item) => {
    const currentItemValue = item[columnName];
    if (!uniqueMap.has(currentItemValue)) {
      columnName === 'category' ? uniqueMap.set(currentItemValue.label) : uniqueMap.set(currentItemValue);
    }
  });
}

/**
 * noramlizeDetailsTableData
 *
 * used to build out a invoice details table data
 *
 * @param {Array} draftInvoiceList - Array of invoices
 * @returns Array of invoice details built for table view
 */
export function noramlizeDetailsTableData(draftInvoiceList = []) {
  const formatted = draftInvoiceList.map((invoice) => {
    const items = invoice.lineItems || [];

    const total = items.reduce((acc, el) => {
      if (el?.payment) {
        acc += Number(el?.payment);
      }

      return acc;
    }, 0);

    const uniqueCategories = new Map();
    const uniquePaymentMethods = new Map();

    populateMap(items, "category", uniqueCategories);
    populateMap(items, "payment_method", uniquePaymentMethods);

    return {
      category: Array.from(uniqueCategories.keys()).join(" / "),
      invoiceStatus: invoice?.invoiceStatus || "",
      start_date: invoice?.start_date,
      end_date: invoice?.end_date,
      total,
      payment_method: Array.from(uniquePaymentMethods.keys()).join(" / "),
      updatedOn: invoice?.updatedOn,
    };
  });

  return formatted;
}

/**
 * normalizeInvoiceItemTypeChartDataset
 *
 * used to build out a item type chart from the provided dataset
 *
 * @param {Array} draftInvoiceList - Array of invoices
 * @returns Object containing the built dataset based on the passed in args
 */
export function normalizeInvoiceItemTypeChartDataset(draftInvoiceList = []) {
  const itemCountMap = {};

  const filteredDraftInvoiceList = draftInvoiceList.filter(Boolean); // remove unwanted values
  if (filteredDraftInvoiceList.length > 0) {
    filteredDraftInvoiceList.forEach(({ lineItems = [] }) => {
      lineItems.forEach((item) => {
        const itemDescription = item?.category?.label || "Unknown Item";
        itemCountMap[itemDescription] =
          (itemCountMap[itemDescription] || 0) + 1;
      });
    });
  }

  const labels = Object.keys(itemCountMap);
  const frequencies = Object.values(itemCountMap);
  const datasets = [
    {
      label: "Item Type Frequency",
      data: frequencies,
      backgroundColor: "rgba(153, 102, 255, 0.7)",
      borderColor: "rgba(153, 102, 255, 1)",
      borderWidth: 1,
    },
  ];

  return {
    labels,
    datasets,
  };
}

/**
 * normalizeInvoiceTrendsChartsDataset
 *
 * used to build out a bar chart dataset for chartjs.
 *
 * @param {Array} draftInvoiceList - Array of invoices
 * @param {String} chartType - The type of chart
 * @returns Object containing the built dataset based on the passed in args
 */
export function normalizeInvoiceTrendsChartsDataset(
  draftInvoiceList = [],
  chartType = "",
) {
  const monthMap = new Map();

  const filteredDraftInvoiceList = draftInvoiceList.filter(Boolean); // remove unwanted values

  filteredDraftInvoiceList.forEach((invoice) => {
    const month = dayjs(invoice.start_date).format("MMMM");
    const taxRate = Number(invoice.tax_rate || 0);

    const collectedTotal = invoice?.lineItems?.reduce((acc, item) => {
      return acc + Number(item?.payment || 0);
    }, 0);

    const taxCollected = parseFloat(
      ((collectedTotal * taxRate) / 100).toFixed(2),
    );

    if (!monthMap.has(month)) {
      monthMap.set(month, {
        collected: 0,
        tax: 0,
      });
    }

    const current = monthMap.get(month);
    monthMap.set(month, {
      collected: current.collected + collectedTotal,
      tax: current.tax + taxCollected,
    });
  });

  const labels = Array.from(monthMap.keys());
  const collectedData = Array.from(monthMap.values()).map(
    (val) => val.collected,
  );
  const taxData = Array.from(monthMap.values()).map((val) => val.tax);

  return [
    {
      labels,
      datasets: [
        {
          label: "Collected Invoice",
          data: collectedData,
          backgroundColor: "rgba(54, 162, 235, 0.7)",
          borderColor: "rgba(54, 162, 235, 1)",
          fill: chartType === "line",
          tension: 0.4,
        },
        {
          label: "Tax Collected",
          data: taxData,
          backgroundColor: "rgba(255, 99, 132, 0.7)",
          borderColor: "rgba(255, 99, 132, 1)",
          fill: chartType === "line",
          tension: 0.4,
        },
      ],
    },
  ];
}

/**
 * normalizeInvoiceTimelineChartDataset
 *
 * used to build out a bar chart dataset for chartjs.
 *
 * @param {Array} draftInvoiceList - Array of invoices
 * @returns Object containing the built dataset based on the passed in args
 */
export function normalizeInvoiceTimelineChartDataset(draftInvoiceList = []) {
  const months = new Set();

  const filteredDraftInvoiceList = draftInvoiceList.filter(Boolean); // remove unwanted values

  filteredDraftInvoiceList.forEach((invoice) => {
    const month = dayjs(invoice.start_date).format("MMMM");
    months.add(month);
  });

  const monthLabels = Array.from(months);

  const datasets = draftInvoiceList.map((invoice, id) => {
    const month = dayjs(invoice.start_date).format("MMMM");
    const index = monthLabels.indexOf(month);

    const duration = dayjs(invoice.end_date).diff(invoice.start_date, "day");
    const data = monthLabels.map((_, idx) => (idx === index ? duration : null));

    return {
      label: `Payment: $${invoice.lineItems?.[0]?.payment} via ${invoice.lineItems?.[0]?.payment_method}`,
      data,
      backgroundColor: id % 2 === 0 ? "#4CAF50" : "rgba(255, 99, 132, 0.7)",
      borderWidth: 1,
    };
  });

  return {
    labels: monthLabels,
    datasets,
  };
}
