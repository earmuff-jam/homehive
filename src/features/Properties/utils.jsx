/**
 * Email Validators
 */
const emailValidators = [
  {
    validate: (value) => value.trim().length <= 0,
    message: "Email address is required",
  },
  {
    validate: (value) => value.trim().length >= 150,
    message: "Email address should be less than 150 characters",
  },
  {
    validate: (value) => !/^[\w-.]+@([\w-]+\.)+[\w-]{2,}$/.test(value),
    message: "Email address is not valid",
  },
];

/**
 * isValid ...
 *
 * function used to determine if an email is valid or not
 *
 * @param {string} email
 * @returns boolean - true / false
 */
export const isValid = (email) => {
  for (const validator of emailValidators) {
    if (validator.validate(email)) {
      return false;
    }
  }
  return true;
};

/**
 * fetchLoggedInUser ...
 *
 * used to retrieve the logged in userId.
 *
 * @returns string - the logged in userId
 */
export const fetchLoggedInUser = () => {
  return JSON.parse(localStorage.getItem("user"));
};
