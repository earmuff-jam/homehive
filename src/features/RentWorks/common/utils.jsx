/**
 * Utility file for properties
 */
import dayjs from "dayjs";

import { getAuth, signOut } from "firebase/auth";
import { authenticatorConfig } from "src/config";

// enum values
export const PaidRentStatusEnumValue = "paid";

/**
 * Email Validators
 */
const emailValidators = [
  {
    validate: (value) => value.trim().length <= 0,
    message: "Email address is required",
  },
  {
    validate: (value) => value.trim().length >= 150,
    message: "Email address should be less than 150 characters",
  },
  {
    validate: (value) => !/^[\w-.]+@([\w-]+\.)+[\w-]{2,}$/.test(value),
    message: "Email address is not valid",
  },
];

/**
 * isValid ...
 *
 * function used to determine if an email is valid or not
 *
 * @param {string} email
 * @returns boolean - true / false
 */
export const isValid = (email) => {
  for (const validator of emailValidators) {
    if (validator.validate(email)) {
      return false;
    }
  }
  return true;
};

/**
 * fetchLoggedInUser ...
 *
 * used to retrieve the logged in userId.
 *
 * @returns string - the logged in userId
 */
export const fetchLoggedInUser = () => {
  return JSON.parse(localStorage.getItem("user"));
};

/**
 * logoutUser ...
 *
 * used to log the user out of the system. function is
 * asyncronous in nature to ensure that the logout is
 * succesful
 *
 * @returns string - the logged in userId
 */
export const logoutUser = async () => {
  const auth = getAuth(authenticatorConfig);
  try {
    await signOut(auth);
    localStorage.removeItem("user");
  } catch (error) {
    /* eslint-disable no-console */
    console.error("Error signing out:", error);
  }
};

/**
 * updateDateTime function
 *
 * util function used to updateDateTime. used to update the
 * next projected rent due date
 *
 * @param {string} startDate - the start date of the event
 * @returns updatedDateTime with a month added.
 */
export const updateDateTime = (startDate) => {
  const today = dayjs();
  const monthsSinceStart = today.diff(startDate, "month");
  const nextDueDate = startDate.add(monthsSinceStart + 1, "month");
  return dayjs(nextDueDate).toISOString();
};

/**
 * formatCurrency ...
 *
 * used to format the current passed in amount.
 *
 * @param {Number} amount - the amount that needs to be formatted, default 0
 *
 * @returns {Number} formatted result
 */
export const formatCurrency = (amt = 0) => {
  return `$${parseInt(amt).toLocaleString()}`;
};

/**
 * derieveTotalRent
 *
 * function used to retrieve the total rent of any given property. For homes
 * with a SoR, rent are calculated per room. The property unit as a whole can have
 * additional charges.
 *
 * @param {Object} property - the property object
 * @param {Array} tenants - the array of tenants that are residing in the property
 * @param {Boolean} isAnyTenantSoR - true / false - determine if the property is single occupancy or not
 *
 * @returns {Number} - amount of rent in US Dollars
 */
export const derieveTotalRent = (property, tenants, isAnyTenantSoR) => {
  const totalRent = Number(property?.rent) + Number(property?.additional_rent); // can have additional charges

  if (isAnyTenantSoR) {
    return tenants.reduce(
      (total, tenant) =>
        total +
        parseInt(tenant.rent || 0) +
        parseInt(property?.additional_rent),
      0,
    );
  } else {
    return totalRent || 0;
  }
};

/**
 * getOccupancyRate ...
 *
 * function used to determine the occupancy rate of the selected
 * property.
 *
 * @param {Object} property - the property object
 * @param {Array} tenants - the array of tenants that are residing in the property
 * @param {Boolean} isAnyTenantSoR - true / false - determine if the property is single occupancy or not
 *
 * @returns {Number} - the percent of the occupancy of the selected property
 */
export const getOccupancyRate = (property, tenants, isAnyTenantSoR) => {
  if (isAnyTenantSoR) {
    const totalUnits = parseInt(property?.units || 0);
    const occupiedUnits = tenants.length;
    return totalUnits > 0 ? Math.round((occupiedUnits / totalUnits) * 100) : 0;
  } else {
    // if !SoR, all tenants count as 1 household member. hence 100% occupancy rate
    return tenants?.length > 0 ? 100 : 0;
  }
};

/**
 * getNextMonthlyDueDate ...
 *
 *
 * function used to return next due date (monthly) based on the original lease start date.
 *
 * @param {string | Date} startDate - The tenant's lease start date.
 * @returns {string} - The next due date in YYYY-MM-DD format.
 */
export function getNextMonthlyDueDate(startDate) {
  if (!startDate) return "";

  const original = dayjs(startDate);
  const today = dayjs();
  const targetDay = original.date();

  const nextDue =
    today.date() <= targetDay
      ? today.set("date", targetDay)
      : today.add(1, "month").set("date", targetDay);

  return nextDue.format("YYYY-MM-DD");
}

/**
 * Checks if rent is currently due.
 *
 * @param {string} startDate - The lease start date in MM-DD-YYYY format.
 * @param {number} gracePeriodDays - Number of grace days before rent is due each month.
 * @param {Array} currentMonthRent - Rent details if exists, for the current month.
 *
 * @returns {boolean} - True if rent is currently due.
 */

export const isRentDue = (startDate, gracePeriod = 3, currentMonthRent) => {
  const today = dayjs();
  const leaseStart = dayjs(startDate, "MM-DD-YYYY");

  if (today.isBefore(leaseStart, "day")) return false;

  const graceDate = today.startOf("month").add(gracePeriod, "day");
  const pastGracePeriod = today.isAfter(graceDate, "day");

  const currentMonth = today.format("MMMM");
  const rentPaid =
    currentMonthRent?.rentMonth === currentMonth &&
    currentMonthRent.status?.toLowerCase() === "paid";
  return pastGracePeriod && !rentPaid;
};

/**
 * getRentStatus ...
 *
 * function used to get the rent status
 * @param {Object} { isPaid, isLate } - object containing these values
 * @returns Object containing the color and label. Eg, { color: "warning", label: "Unpaid" }
 */
export function getRentStatus({ isPaid, isLate }) {
  if (isPaid) return { color: "success", label: "Paid" };
  if (isLate) return { color: "error", label: "Overdue" };
  return { color: "warning", label: "Unpaid" };
}

/**
 * getCurrentMonthPaidRent ...
 *
 * fn used to get the current month rent.
 *
 * @param {Array} allRents - list of all the rents for the given property
 * @returns {Object} - current month rent
 */
export function getCurrentMonthPaidRent(allRents = []) {
  const currentMonth = dayjs().format("MMMM"); // e.g. "July"

  return allRents.find(
    (rent) =>
      rent.rentMonth === currentMonth && rent.status?.toLowerCase() === "paid",
  );
}
