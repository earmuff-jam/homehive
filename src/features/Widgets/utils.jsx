import dayjs from "dayjs";

/**
 * fakeDataset
 *
 * creates a fake dataset function for testing purposes only.
 * TODO: REMOVE THIS AFTER DASHBOARD WORK IS COMPLETED.
 */
export function fakeDataset() {
  const draftData = {
    title: "Rent for the month of March",
    caption: "Test title",
    note: "No additional notes",
    start_date: "04-01-2025",
    end_date: "04-30-2025",
    tax_rate: "1",
    invoice_header: "test",
    items: [
      {
        descpription: "test",
        caption: "test",
        quantity: "1",
        price: "1389.98",
        payment: "1389.98",
        payment_method: "zelle",
      },
    ],
    updated_on: "2025-04-15T13:14:09.627Z",
  };

  const draftData2 = {
    title: "Rent for the month of April",
    caption: "Test title",
    note: "No additional notes",
    start_date: "05-01-2025",
    end_date: "05-31-2025",
    tax_rate: "1",
    invoice_header: "test",
    items: [
      {
        descpription: "test",
        caption: "test",
        quantity: "1",
        price: "1389.98",
        payment: "1389.98",
        payment_method: "zelle",
      },
    ],
    updated_on: "2025-04-15T13:14:09.627Z",
  };

  localStorage.setItem(
    "invoiceDataList",
    JSON.stringify([draftData, draftData2])
  );
}

/**
 * normalizeBarChartDataset
 *
 * used to build out a bar chart dataset for chartjs.
 *
 * @param {Array} draftInvoiceList - Array of invoices
 *
 * @returns Object containing single dataset
 */
export function normalizeDataset(draftInvoiceList = []) {
  const months = new Set();

  draftInvoiceList.forEach((invoice) => {
    const month = dayjs(invoice.start_date).format("MMMM");
    months.add(month);
  });

  const monthLabels = Array.from(months);

  const datasets = draftInvoiceList.map((invoice) => {
    const month = dayjs(invoice.start_date).format("MMMM");
    const index = monthLabels.indexOf(month);

    const duration = dayjs(invoice.end_date).diff(invoice.start_date, "day");
    const data = monthLabels.map((_, idx) => (idx === index ? duration : null));

    return {
      label: `Payment: $${invoice.items?.[0]?.payment} via ${invoice.items?.[0]?.payment_method}`,
      data,
      backgroundColor: "#4CAF50",
      borderWidth: 1,
    };
  });

  return {
    labels: monthLabels,
    datasets,
  };
}
