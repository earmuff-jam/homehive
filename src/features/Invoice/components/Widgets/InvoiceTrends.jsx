import React, { useEffect, useState } from "react";

import { Bar, Line } from "react-chartjs-2";

import { BarChartRounded, StackedLineChartRounded } from "@mui/icons-material";
import { Box, Stack, ToggleButton, ToggleButtonGroup } from "@mui/material";
import {
  BarElement,
  CategoryScale,
  Chart as ChartJS,
  Legend,
  LineElement,
  LinearScale,
  PointElement,
  Title,
  Tooltip,
} from "chart.js";
import EmptyComponent from "common/EmptyComponent";
import RowHeader from "common/RowHeader/RowHeader";
import { normalizeInvoiceTrendsChartsDataset } from "features/Invoice/components/Widgets/utils";

ChartJS.register(
  CategoryScale,
  LinearScale,
  BarElement,
  LineElement,
  PointElement,
  Tooltip,
  Legend,
  Title,
);

const InvoiceTrendsChart = ({ label, caption }) => {
  const [chartData, setChartData] = useState(null);
  const [chartType, setChartType] = useState("bar"); // or "line"

  const handleChartType = (ev, draftChartType) => {
    if (draftChartType !== null) {
      setChartType(draftChartType);
    }
  };

  useEffect(() => {
    const draftData = JSON.parse(localStorage.getItem("pdfDetails"));
    if (draftData) {
      const chartData = normalizeInvoiceTrendsChartsDataset(
        [draftData],
        chartType,
      );
      setChartData(chartData[0]);
    }
  }, [chartType]);

  const options = {
    responsive: true,
    plugins: {
      title: {
        display: true,
        text: "Invoice Totals & Tax Collected Over Time",
      },
      legend: {
        position: "top",
      },
    },
    scales: {
      y: {
        beginAtZero: true,
        stacked: chartType === "bar",
      },
      x: {
        stacked: chartType === "bar",
      },
    },
  };

  return (
    <Stack data-tour={"dashboard-5"}>
      <Stack direction="row" justifyContent="space-between">
        <RowHeader
          title={label}
          caption={caption}
          sxProps={{
            textAlign: "left",
            fontWeight: "bold",
            color: "text.secondary",
          }}
        />

        <Box>
          <ToggleButtonGroup
            value={chartType}
            exclusive
            onChange={handleChartType}
            aria-label="bar or line chart"
          >
            <ToggleButton value="bar" aria-label="bar chart" size="small">
              <BarChartRounded />
            </ToggleButton>
            <ToggleButton value="line" aria-label="line chart" size="small">
              <StackedLineChartRounded />
            </ToggleButton>
          </ToggleButtonGroup>
        </Box>
      </Stack>

      <Box>
        {chartData === null ? (
          <EmptyComponent />
        ) : chartType === "bar" ? (
          <Bar data={chartData} options={options} />
        ) : (
          <Line data={chartData} options={options} />
        )}
      </Box>
    </Stack>
  );
};

export default InvoiceTrendsChart;
