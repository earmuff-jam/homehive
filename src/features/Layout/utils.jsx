import { PropertyRouteUri } from "common/utils";

/**
 * retreiveTourKey ...
 *
 * used to return the correct key for tour steps.
 *
 * @param {string} currentUri - the string representation of the current uri
 * @param {string} expectedStrValue - the expected string value, if matched manipulates data.
 *
 * @returns {string} - the key to retrieve the tour from
 */
export const retrieveTourKey = (currentUri, expectedStrValue) => {
  const isDynamicPropertyPage =
    currentUri.includes(`/${expectedStrValue}/`) &&
    currentUri.split("/")[2] === expectedStrValue;

  // individual properties can share the same help && support
  return isDynamicPropertyPage ? PropertyRouteUri : currentUri;
};
