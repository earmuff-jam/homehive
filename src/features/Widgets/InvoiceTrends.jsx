import { useEffect, useState } from "react";
import { Bar, Line } from "react-chartjs-2";
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  BarElement,
  LineElement,
  PointElement,
  Tooltip,
  Legend,
  Title,
} from "chart.js";
import { Box, Stack, ToggleButton, ToggleButtonGroup } from "@mui/material";
import RowHeader from "src/common/RowHeader/RowHeader";
import { BarChartRounded, StackedLineChartRounded } from "@mui/icons-material";
import { normalizeInvoiceTrendsChartsDataset } from "src/features/Widgets/utils";
import EmptyComponent from "src/features/Widgets/EmptyComponent";

ChartJS.register(
  CategoryScale,
  LinearScale,
  BarElement,
  LineElement,
  PointElement,
  Tooltip,
  Legend,
  Title
);

const InvoiceTrendsChart = ({ label, caption }) => {
  const [chartData, setChartData] = useState(null);
  const [chartType, setChartType] = useState("bar"); // or "line"

  const handleChartType = (ev, draftChartType) => {
    if (draftChartType !== null) {
      setChartType(draftChartType);
    }
  };

  useEffect(() => {
    const draftData = JSON.parse(localStorage.getItem("pdfDetails"));
    const draftDataList = [draftData];
    if (Array.isArray(draftDataList) && draftDataList.length > 0) {
      const chartData = normalizeInvoiceTrendsChartsDataset(
        draftDataList,
        chartType
      );

      setChartData(chartData[0]);
    }
  }, [chartType]);

  const options = {
    responsive: true,
    plugins: {
      title: {
        display: true,
        text: "Invoice Totals & Tax Collected Over Time",
      },
      legend: {
        position: "top",
      },
    },
    scales: {
      y: {
        beginAtZero: true,
        stacked: chartType === "bar",
      },
      x: {
        stacked: chartType === "bar",
      },
    },
  };

  return (
    <Stack>
      <Stack direction="row" justifyContent="space-between">
        <RowHeader
          title={label}
          caption={caption}
          sxProps={{
            textAlign: "left",
            fontWeight: "bold",
            color: "text.secondary",
          }}
        />

        <Box>
          <ToggleButtonGroup
            value={chartType}
            exclusive
            onChange={handleChartType}
            aria-label="bar or line chart"
          >
            <ToggleButton value="bar" aria-label="bar chart" size="small">
              <BarChartRounded />
            </ToggleButton>
            <ToggleButton value="line" aria-label="line chart" size="small">
              <StackedLineChartRounded />
            </ToggleButton>
          </ToggleButtonGroup>
        </Box>
      </Stack>

      <Box>
        {chartData === null ? (
          <EmptyComponent />
        ) : chartType === "bar" ? (
          <Bar data={chartData} options={options} />
        ) : (
          <Line data={chartData} options={options} />
        )}
      </Box>
    </Stack>
  );
};

export default InvoiceTrendsChart;
