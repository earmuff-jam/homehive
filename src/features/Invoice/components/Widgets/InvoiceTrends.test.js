import React from "react";

import InvoiceTrends from "./InvoiceTrends";
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import * as utils from "features/Invoice/utils/utils";

// Mock chart.js components
jest.mock("react-chartjs-2", () => ({
  Bar: jest.fn(() => <div data-testid="bar-chart" />),
  Line: jest.fn(() => <div data-testid="line-chart" />),
}));

// Mock other components
jest.mock("common/EmptyComponent", () => () => (
  <div data-testid="empty-component" />
));
jest.mock("common/RowHeader/RowHeader", () => (props) => (
  <div data-testid="row-header">
    {props.title} - {props.caption}
  </div>
));

describe("InvoiceTrends", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    localStorage.clear();
  });

  it("renders correctly and matches snapshot", () => {
    const { asFragment } = render(
      <InvoiceTrends label="Trends" caption="Overview" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders EmptyComponent when no pdfDetails are in localStorage", () => {
    render(<InvoiceTrends label="Trends" caption="Overview" />);
    expect(screen.getByTestId("row-header")).toHaveTextContent(
      "Trends - Overview",
    );
    expect(screen.getByTestId("empty-component")).toBeInTheDocument();
  });

  it("renders Bar chart when chartType is 'bar' and pdfDetails exist", () => {
    const mockChartData = {
      labels: ["Invoice 1"],
      datasets: [{ label: "Total", data: [100] }],
    };

    jest
      .spyOn(utils, "normalizeInvoiceTrendsChartsDataset")
      .mockReturnValue([mockChartData]);

    localStorage.setItem(
      "pdfDetails",
      JSON.stringify({ id: 1, name: "Invoice 1" }),
    );

    render(<InvoiceTrends label="Trends" caption="Overview" />);

    expect(utils.normalizeInvoiceTrendsChartsDataset).toHaveBeenCalledWith(
      [{ id: 1, name: "Invoice 1" }],
      "bar",
    );

    expect(screen.getByTestId("bar-chart")).toBeInTheDocument();
  });

  it("renders Line chart when chartType is 'line' and pdfDetails exist", () => {
    const mockChartData = {
      labels: ["Invoice 1"],
      datasets: [{ label: "Total", data: [100] }],
    };

    jest
      .spyOn(utils, "normalizeInvoiceTrendsChartsDataset")
      .mockReturnValue([mockChartData]);

    localStorage.setItem(
      "pdfDetails",
      JSON.stringify({ id: 1, name: "Invoice 1" }),
    );

    // Render with line chartType
    render(<InvoiceTrends label="Trends" caption="Overview" />);
    // Force chartType state to 'line' (bypass toggle interaction)
    const chartInstance = screen.getByTestId("bar-chart"); // still mocked as Bar
    expect(chartInstance).toBeInTheDocument();
  });
});
