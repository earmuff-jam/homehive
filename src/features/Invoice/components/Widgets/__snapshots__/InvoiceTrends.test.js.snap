// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvoiceTrends renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-nen11g-MuiStack-root"
    data-tour="dashboard-5"
  >
    <div
      class="MuiStack-root css-gmwslw-MuiStack-root"
    >
      <div
        data-testid="row-header"
      >
        Trends - Overview
      </div>
      <div
        class="MuiBox-root css-0"
      >
        <div
          aria-label="bar or line chart"
          class="MuiToggleButtonGroup-root MuiToggleButtonGroup-horizontal css-1qb8o-MuiToggleButtonGroup-root"
          role="group"
        >
          <button
            aria-label="bar chart"
            aria-pressed="true"
            class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root Mui-selected MuiToggleButton-sizeSmall MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-firstButton css-qofxz1-MuiButtonBase-root-MuiToggleButton-root"
            tabindex="0"
            type="button"
            value="bar"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="BarChartRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 20c1.1 0 2-.9 2-2v-7c0-1.1-.9-2-2-2s-2 .9-2 2v7c0 1.1.9 2 2 2m10-5v3c0 1.1.9 2 2 2s2-.9 2-2v-3c0-1.1-.9-2-2-2s-2 .9-2 2m-4 5c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2s-2 .9-2 2v12c0 1.1.9 2 2 2"
              />
            </svg>
          </button>
          <button
            aria-label="line chart"
            aria-pressed="false"
            class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-sizeSmall MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-lastButton css-qofxz1-MuiButtonBase-root-MuiToggleButton-root"
            tabindex="0"
            type="button"
            value="line"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="StackedLineChartRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m2.79 14.78-.09-.09a.996.996 0 0 1 0-1.41l6.09-6.1c.39-.39 1.02-.39 1.41 0l3.29 3.29 6.39-7.18c.38-.43 1.05-.44 1.45-.04.37.38.39.98.04 1.37l-7.17 8.07c-.38.43-1.04.45-1.45.04L9.5 9.48l-5.3 5.3c-.38.39-1.02.39-1.41 0m1.41 6 5.3-5.3 3.25 3.25c.41.41 1.07.39 1.45-.04l7.17-8.07c.35-.39.33-.99-.04-1.37-.4-.4-1.07-.39-1.45.04l-6.39 7.18-3.29-3.29a.996.996 0 0 0-1.41 0l-6.09 6.1c-.39.39-.39 1.02 0 1.41l.09.09c.39.39 1.03.39 1.41 0"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        data-testid="empty-component"
      />
    </div>
  </div>
</DocumentFragment>
`;
